### PR TITLE
Qg path sampler

### PR DESCRIPTION
**Pull request by riyazbhat**
# Pull Request      

## What does this PR do?

Closes [#133](https://github.com/primeqa/primeqa/issues/133)
- Extends QG over Hybrid context i.e., tables + linked text
   * Adds [Path Sampler](https://github.com/primeqa/primeqa/blob/qg_path-sampler/primeqa/qg/models/hybrid_qg/path_sampler.py) for QG inference
   * Adds [Processor](https://github.com/primeqa/primeqa/blob/qg_path-sampler/primeqa/qg/processors/hybrid_qg/hybridqa_processor.py) for QG training over [Hybrid QA](https://huggingface.co/datasets/hybrid_qa) dataset
      - Performance of the trained model on the Dev set:
      
      	|Rouge 1 | Rouge 2 | Rouge L | Rouge LSum|
		|-----|-----------|--------------|------|
		|0.55|0.36|0.51|0.51|
      
   * Removes NLTK and Spacy dependencies

Notes:
- Replace `(issue)` **above** ↑↑↑ with the issue this PR closes to automatically link the two.
This must be done when the PR is created.
- Add multiple `Closes #(issue)` as needed.
- If this PR is work towards but does not close an issue, simply tag the issue without mentioning `Closes`.

## Description

**Describe** the changes proposed by this PR below to give the reviewer context below ↓↓↓

```Extends QG over Table+Text```     

### Request Review

Be sure to **request a review** from one or more reviewers (unless the PR is to an unprotected branch).

## Versioning

When opening a PR to make changes to PrimeQA (i.e. `primeqa/`) master, be sure to increment the version following
[semantic versioning](https://semver.org/).  The VERSION is stored [here](https://github.com/primeqa/primeqa/blob/main/VERSION) 
and is incremented using `bump2version {patch,minor,major}` as described in the (development guide documentation)[development.html].

 - [ ] Have you updated the VERSION?
 - [ ] Or does this PR not change the `primeqa` package or was not into master?

After pulling in changes from master to an existing PR, ensure the VERSION is updated appropriately.
This may require bumping the version again if it has been previously bumped.

If you're not quite ready yet to post a PR for review, feel free to open a [draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

### After Merging
If merging into master and VERSION was updated, after this PR is merged:
- [ ] [Create a release](https://docs.github.com/en/github/administering-a-repository/releasing-projects-on-github/managing-releases-in-a-repository#creating-a-release) from the master with version equal to [VERSION](https://github.com/primeqa/primeqa/blob/main/VERSION)
- [ ] Not merging to master or VERSION not updated

### Checklist

Review the following and mark as completed:
- [x] [Tag an issue or issues](#what-does-this-pr-do) this PR addresses.
- [x] Added [description](#description) of changes proposed.
- [x] Review requested as [appropriate](#request-review).
- [ ] Version bumped as [appropriate](#Versioning).
- [x] New classes, methods, and functions documented.
- [x] Documentation for modified code is updated.
- [ ] Built documentation to confirm it renders as expected (see [here](development.md)).
- [x] Code cleaned up and commented out code removed.
- [x] Tests added to ensure all functionalities tested at >= 60% unit test coverage (see [here](development.md)).
- [x] Code cleaned up and commented out code removed.
- [ ] Release created [as needed](#after-merging) after merging.

### CODE REVIEW

**Comment by jsmcibm** [comment]
I think it would greatly clarify the overall purpose of this PR if you could
(1) confirm that hybridqg_inference.ipynb is the place to start reading about it.
(2) add a few extra sentences at the top and bottom explaining the purpose, and why the generated questions are correct
(3) make the passages wrapped ( I can't seem to scroll right in the browser to read the entire passage

**Comment by riyazbhat** [comment]
> I think it would greatly clarify the overall purpose of this PR if you could (1) confirm that hybridqg_inference.ipynb is the place to start reading about it. (2) add a few extra sentences at the top and bottom explaining the purpose, and why the generated questions are correct (3) make the passages wrapped ( I can't seem to scroll right in the browser to read the entire passage

1) Yes, starting points should be to go throw hybridqg_inference.ipynb and hybridqg_training.ipynb notebooks.
2) Adding the performance scores to validate the quality of generated questions.
3) Done.

File: primeqa/qg/models/hybrid_qg/path_sampler.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+from sklearn.metrics.pairwise import linear_kernel
+from sklearn.feature_extraction.text import TfidfVectorizer
+
+import stanza
+
+from primeqa.qg.utils.constants import QGSpecialTokens
+
+
+class PathSampler():
```
Comment (seirasto) [inline primeqa/qg/models/hybrid_qg/path_sampler.py:12]: Can you please add some more documentation here and perhaps also add an example of what the chains look like?

File: primeqa/qg/models/qg_model.py [hunk 2]
```diff
@@ -52,33 +61,110 @@ def tokenizer(self):
...
+                doc = self.path_sampler.nlp_model(question)
+                hallucinated = set([])
+                for entity in doc.sentences[0].entities:
+                    entity = entity.text.lower()
+                    entity_tokens = set([etok.text.strip().lower() for etok in self.path_sampler.nlp_model.process(\
+                                        entity, processors='tokenize').sentences[0].tokens if etok.text.strip().lower() != "'s"])
+                    has_qs_words = question_words.intersection(entity_tokens)
+                    if has_qs_words: continue
+
+                    hcount = 0
```
Comment (seirasto) [inline primeqa/qg/models/qg_model.py:99]: Can you please add a comment here explaining when a token is considered an hallucination

File: primeqa/qg/models/qg_model.py [hunk 3]
```diff
@@ -52,33 +61,110 @@ def tokenizer(self):
...
+                    qdict_copy = copy.deepcopy(qdict)
+                    qdict_copy['question'] = question
+                    if flag:
+                        reserve_non_hallucinated_qdict.append(qdict_copy)
+                    else:
+                        new_qdicts.append(qdict_copy)
+                    flag = True
+
+        if not new_qdicts:
+            return reserve_hallucinated_qdict
```
Comment (seirasto) [inline primeqa/qg/models/qg_model.py:120]: If everything is hallucinated then don't prune?

File: primeqa/qg/models/qg_model.py [hunk 4]
```diff
@@ -52,33 +61,110 @@ def tokenizer(self):
...
+        elif len(new_qdicts) < num_instances:
```
Comment (seirasto) [inline primeqa/qg/models/qg_model.py:121]: What is happening here?

File: primeqa/qg/models/qg_model.py [hunk 5]
```diff
@@ -52,33 +61,110 @@ def tokenizer(self):
...
+                        num_questions_per_instance, answers_list, id_list)
+        else:
+            input_str_list, answer_list, id_question_list = self.path_sampler.create_qg_input(data_list, \
+                        num_questions_per_instance, id_list)
 
         input_ids = self._tokenizer(input_str_list, 
-            return_tensors='pt', 
-            padding=True,
-            truncation=True).input_ids
+                return_tensors='pt',
```
Comment (seirasto) [inline primeqa/qg/models/qg_model.py:154]: This if is indented extra tab

File: notebooks/qg/hybridqg_inference.ipynb [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "984e88f0",
+   "metadata": {},
+   "source": [
+    "<br>\n",
+    "\n",
+    "## Sample table\n",
+    "\n",
+    "Hybrid context should be passed as a `list` of `tuples`. Each `tuple` will correspond to a table (as a `dict` with keys `\"header\"` and `\"rows\"`) and linked passages (as a `dict`)."
```
Comment (seirasto) [inline notebooks/qg/hybridqg_inference.ipynb:201]:
```suggestion
    "The hybrid context should be passed as a `list` of `tuples`. Each `tuple` will correspond to a table (as a `dict` with keys `\"header\"` and `\"rows\"`) and linked passages (as a `dict`)."
```

File: notebooks/qg/hybridqg_inference.ipynb [hunk 7]
```diff
@@ -0,0 +1,377 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "489bc0df",
+   "metadata": {},
+   "source": [
+    "# Multihop Hybrid (Table + text) Question Generation: Inference example\n",
+    "\n",
+    "In this notebook, we will show how to use a pretrained HybridQG model to generate questions. We first sample hybrid chains from a given table and its linked passages, and then use a text-to-text transformer (T5) to transcribe the chains to a natural language question. "
```
Comment (seirasto) [inline notebooks/qg/hybridqg_inference.ipynb:10]: Can you show intermediate output in the notebook, for example what the chains look like?

File: primeqa/qg/processors/hybrid_qg/hybridqa_processor.py [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+        """
+        self.tokenizer = tokenizer
+        self.input_max_len = input_max_len
+        self.target_max_len = target_max_len
+        self.nlp_model = stanza.Pipeline(lang='en', processors='tokenize', verbose=False)
+        self.answer_separator = " %s " % (QGSpecialTokens.ans)
+        self.hop_separator = " %s " % (QGSpecialTokens.hsep)
+        self.meta_separator = " %s " % (QGSpecialTokens.header)
+
+    def __call__(self, dataset) -> Dataset:
```
Comment (seirasto) [inline primeqa/qg/processors/hybrid_qg/hybridqa_processor.py:34]: Can you add more docstring comments to this file? Also in some cases there are comments but they are not docstrings

File: notebooks/qg/hybridqg_training.ipynb [hunk 9]
```diff
@@ -0,0 +1,1032 @@
...
+    "training_args.prediction_loss_only = False"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "---\n",
+    "## HybridQA data\n",
+    "Here we load one instance of HybridQA and visualize it. <font color='red'>This part of the code is not needed to train the model </font>"
```
Comment (seirasto) [inline notebooks/qg/hybridqg_training.ipynb:79]: I think it would be helpful if you could also show this instance using a table.

File: notebooks/qg/hybridqg_training.ipynb [hunk 10]
```diff
@@ -0,0 +1,1032 @@
...
+    "print(train_results.metrics)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Evaluation\n",
+    "\n",
+    "Here we evaluate the trained model on validation set"
```
Comment (seirasto) [inline notebooks/qg/hybridqg_training.ipynb:947]:
```suggestion
    "Here we evaluate the trained model on the validation set"
```

File: notebooks/qg/hybridqg_inference.ipynb [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "984e88f0",
+   "metadata": {},
+   "source": [
+    "<br>\n",
+    "\n",
+    "## Sample table\n",
```
Comment (seirasto) [inline notebooks/qg/hybridqg_inference.ipynb:199]: Can you show this table in table form as well? (e.g. an image of the table)

**Review by seirasto — COMMENTED** [review]
Looks good for the most part. I've left some comments. Does there need to be two notebooks for the two inference examples (hybrid and passage) or can they be in the same notebook? Will there be updates to the READMEs for these updates?

File: primeqa/qg/models/hybrid_qg/path_sampler.py (hunk 1 above)
Comment (riyazbhat) [reply primeqa/qg/models/hybrid_qg/path_sampler.py:12]: I have add some more documentation to this file and pushed the changes.

File: primeqa/qg/models/qg_model.py (hunk 2 above)
Comment (riyazbhat) [reply primeqa/qg/models/qg_model.py:99]: Yes, I have added a definition of hallucination in the function.

File: primeqa/qg/models/qg_model.py (hunk 3 above)
Comment (riyazbhat) [reply primeqa/qg/models/qg_model.py:120]: Yes, just return some randomly sampled question from hallucinated questions, so that we don't return an empty list/dict.

File: primeqa/qg/models/qg_model.py (hunk 4 above)
Comment (riyazbhat) [reply primeqa/qg/models/qg_model.py:121]: In case most of the generated questions are hallucinated, pick some hallucinated questions so that the number of output questions equals `num_instances`.

File: primeqa/qg/processors/hybrid_qg/hybridqa_processor.py (hunk 8 above)
Comment (riyazbhat) [reply primeqa/qg/processors/hybrid_qg/hybridqa_processor.py:34]: Done

File: notebooks/qg/hybridqg_training.ipynb (hunk 9 above)
Comment (riyazbhat) [reply notebooks/qg/hybridqg_training.ipynb:79]: I am not sure if this nested instance can be easily converted to a table.

File: notebooks/qg/hybridqg_inference.ipynb (hunk 11 above)
Comment (riyazbhat) [reply notebooks/qg/hybridqg_inference.ipynb:199]: Do you mean just the table part or both table and text into one single table? The latter might not be possible due to large text passages.

File: notebooks/qg/hybridqg_inference.ipynb (hunk 7 above)
Comment (riyazbhat) [reply notebooks/qg/hybridqg_inference.ipynb:10]: I have added a code snippet for printing hybrid chains.

File: notebooks/qg/hybridqg_inference.ipynb (hunk 11 above)
Comment (seirasto) [reply notebooks/qg/hybridqg_inference.ipynb:106]: I meant just the table part, if this is doable.

File: notebooks/qg/hybridqg_inference.ipynb (hunk 6 above)
Comment (seirasto) [reply notebooks/qg/hybridqg_inference.ipynb:108]: Did you miss this suggestion?

**Review by seirasto — APPROVED** [review]
LGTM! Still a few small comments to address if you can.

File: primeqa/qg/models/qg_model.py (hunk 4 above)
Comment (seirasto) [reply primeqa/qg/models/qg_model.py:121]: Can you add this explanation as a comment?

File: primeqa/qg/models/qg_model.py (hunk 2 above)
Comment (seirasto) [reply primeqa/qg/models/qg_model.py:99]: I don't see this change

File: primeqa/qg/models/qg_model.py (hunk 4 above)
Comment (riyazbhat) [reply primeqa/qg/models/qg_model.py:121]: I have added it as a comment to the code.

File: primeqa/qg/models/qg_model.py (hunk 2 above)
Comment (riyazbhat) [reply primeqa/qg/models/qg_model.py:99]: """
        Our approach to pruning hallucinated questions uses an entity lookup to check that 
        a generated entity or a part of it is present in the hybrid context. If the entity is not present, 
        the question and the entity are considered to be hallucinated.

        Args:
            qdicts (list): A list of dicts, each dict contains, an answer, context and (num_instances) generated questions.
        Returns:
            Updated qdicts without any hallucinated questions. Each dict will have only one question now.
        """
I have pushed the changes with this docstring.

File: notebooks/qg/hybridqg_inference.ipynb (hunk 6 above)
Comment (riyazbhat) [reply notebooks/qg/hybridqg_inference.ipynb:108]: It seems so. I will update notebook with the change.

**Comment by riyazbhat** [comment]
> LGTM! Still a few small comments to address if you can.

Incorporated all of your suggestions, but one. Converting a highly nested data object is very tricky, so I am keeping it as such.

File: setup.py [hunk 12]
```diff
@@ -70,11 +70,10 @@
...
     "pytest-rerunfailures~=10.2": ["tests"],
-    "scikit-learn~=1.0.2": ["install", "gpu"],
-    "signals~=0.0.2": ["install", "gpu"],
-    "spacy~=3.2.2": ["install", "gpu"],
-    "stanza~=1.4.0": ["install", "gpu"],
-    "torch~=1.11.0": ["install", "gpu"],
+    "scikit-learn~=1.0.2": ["install"],
+    "signals~=0.0.2": ["install"],
+    "stanza~=1.4.0":["install"],
+    "torch~=1.11.0": ["install"],
```
[the remark this answers is not in the document]
Comment (riyazbhat) [reply setup.py]: Might have happened by mistake. Gpu is back now for all of these packages.

File: primeqa/qg/metrics/generation_metrics.py [hunk 13]
```diff
@@ -1,23 +1,6 @@
-import nltk
 import numpy as np
 from datasets import load_metric
-from nltk.tokenize import sent_tokenize
 
-def bleuscore(pred_list):
```
[the remark this answers is not in the document]
Comment (riyazbhat) [reply primeqa/qg/metrics/generation_metrics.py:6]: Yes, to remove the nltk dependency. We have ROUGE as a metric for QA anyway.

File: primeqa/ir/dense/colbert_top/conda_env.yml [hunk 14]
```diff
@@ -19,7 +19,6 @@ dependencies:
     - jupyter
     - jupyterlab
     - scipy
-    - spacy
```
[the remark this answers is not in the document]
Comment (riyazbhat) [reply primeqa/ir/dense/colbert_top/conda_env.yml:22]: There is no reference to spacy at all in `ir`. So I guess it should be ok.
